crdt/woot.go

```diff
@@ -256,7 +259,9 @@ func (doc *Document) IntegrateInsert(char, charPrev, charNext Character) (*Docum
 // GenerateInsert generates a character for a given value.
 func (doc *Document) GenerateInsert(position int, value string) (*Document, error) {
 	// Increment local clock.
+	mu.Lock()
 	LocalClock++
+	mu.Unlock()
```

Comment:
Using a mutex here is a bit confusing - what are the performance implications on doing large-scale insertion operations? (pasting large pieces of text)

I'm assuming that it has little to no effect on performance, but it would be great if you could test it out by pasting large pieces of text, and checking if the operations take more time or not.

Reply:
While I do feel that the mutex is necessary, I'm also a bit hesitant about the setup and teardown of the mutex at each insertion, as this increases the duration for an insert operation, and *might* reduce the overall performance of the editor.

Not a deal breaker, as *correctness > performance* in my opinion.

Reply:
Good idea! Benchmarks are fun! Here's the benchmark code:

```
func BenchmarkInsert(b *testing.B) {
	doc := New()
	for n := 0; n < b.N; n++ {
		doc.Insert(n, "a")
	}
}
```

I think this is equivalent to what happens when you copy and paste into Pairpad, but please correct me if I'm wrong.

I ran the benchmarks with both versions of the `Insert` method, with and without the lock. I ran the benchmark with the command `go test -bench=. -benchtime=XXXx` where XXX is the number of times to run the benchmark (you have to include the benchtime flag because the performance never stabilizes, it decreases exponentially with the size of the document). I ran each benchmark command 5 times and discarded the fastest and slowest results, and averaged the remaining 3.

Here are the results: 

**Without lock**
```
XXX       ns/op
10        2886 
100       4500 
1000      45599 
5000      782290
```

**With lock**
```
XXX       ns/op
10        3131
100       4676
1000      46424
5000      781946
```

I couldn't check the difference for 10000 inserts because the benchmark didn't end after 25 seconds! The benchmark for 5000 inserts took about 4 seconds, by comparison. The exponentially worse performance is due to the recursive nature of the `IntegrateInsert` method, right? I don't think we can do much about that short of changing the algorithm.

Anyway, like you said, there's not a huge effect on performance, because [contention is what makes locks expensive](https://preshing.com/20111118/locks-arent-slow-lock-contention-is/). The performance difference between the lock vs lock-less inserts actually decreases with larger documents, probably because the time of each operation starts to be more dominated by the WOOT algorithm itself, so by the time we get to 5000 characters the cost of the lock doesn't show up in the benchmark any more.

Reply:
>  The exponentially worse performance is due to the time complexity of the recursive IntegrateInsert method, right? I don't think we can do much about that short of changing the algorithm.

True. There isn't much room for optimization here. WOOT is fairly old at this point ([the WOOT paper](https://inria.hal.science/inria-00108523/document) was published in 2006), although, I won't deny the fact that it is still a solid choice for collaborative editing. Upgrading to a *performant* algorithm should be straightforward through the CRDT layer - however that isn't a top priority for the project, at least for now.

I'm happy with the benchmarks, and I believe that ~5k insertions being done in <5s is a good foundation for the editor. Thanks for the detailed explanation!